### PR TITLE
Fix pytest warnings (pytest-3.1)

### LINKED ISSUE #4303: pytest-3.1 raises lots of warnings running our tests
Latest `pytest-3.1.x` versions raise several warnings when running our tests, mostly telling that `unittest` functions `assertEquals` and `assertRaisesRegexp` should not be used anymore. We should get rid of those warnings...

### PR DESCRIPTION
**Pull request by thisch**
The recently released pytest-3.1.0 outputs all warnings at the end of
the testrun. This commit fixes all warnings in the coala unit tests.

Closes #4303 

see https://docs.pytest.org/en/latest/warnings.html

**For short term contributors:** we understand that getting your commits well
defined like we require is a hard task and takes some learning. If you
look to help without wanting to contribute long term there's no need
for you to learn this. Just drop us a message and we'll take care of brushing
up your stuff for merge!

### Checklist

- [x] I read the [commit guidelines](http://coala.io/commit) and I've followed
      them.
- [x] I ran coala over my code locally. (*All commits have to pass
      individually.* It is not sufficient to have "fixup commits" on your PR,
      our bot will still report the issues for the previous commit.) You will
      likely receive a lot of bot comments and build failures if coala does not
      pass on every single commit!

After you submit your pull request, **DO NOT click the 'Update Branch' button.**
When asked for a rebase, consult [coala.io/rebase](https://coala.io/rebase)
instead.

Please consider helping us by reviewing other peoples pull requests as well:

- pick up any PR at <https://coala.io/review>
- review it (check <https://coala.io/reviewing> for more info)
- if you are sure that it needs work, use `cobot mark wip <URL>` to get it out
  of the review queue.

The more you review, the more your score will grow at coala.io and we will
review your PRs faster!

### CODE REVIEW

**Comment by supergr35** [comment]
Squash your commits and fix the commit body line error 👍

**Comment by sims1253** [comment]
You can find a tutorial on how to squash your commits [here](http://api.coala.io/en/latest/Developers/Git_Basics.html#squashing-your-commits)

You also have a line in your commit message that is too long. After rebasing (and squashing) you can use `git commit --amend` to alter your current commit message.
Then you have to `git push -f`to force a push, overwriting the current online version with your altered version.

**Comment by thisch** [comment]
If I squash the three commmits into one, which string should I use before the colon?

File: coalib/parsing/Globbing.py [hunk 1]
```diff
@@ -191,7 +191,7 @@ def translate(pattern):
                 regex += '[' + sequence + ']'
         else:
             regex = regex + re.escape(char)
-    return regex + '\\Z(?ms)'
+    return '(?ms)' + regex + '\\Z'
```
Comment (userzimmermann) [inline coalib/parsing/Globbing.py]: AFAIK it doesn't matter where those inline `(?flags)` are located in the regex string

Comment (thisch) [inline coalib/parsing/Globbing.py]: There is a deprecationwarning which states that flags in expression strings should be at the start of the string. 

See 
> Flags should be used first in the expression string. (https://docs.python.org/3/library/re.html)
and https://bugs.python.org/issue22493

**Comment by thisch** [comment]
ping

**Comment by thisch** [comment]
There is no corresponding issue that I'm solving. I just tried to run the tests with the latest pytest release and found a bunch of warnings.

**Comment by userzimmermann** [comment]
@thisch Of course there is an issue for your PR ;) https://github.com/coala/coala/issues/4303

And @jayvdb is right: Please add a little more detail to your commit message body. Otherwise everything is very fine with your PR! :)

**Comment by thisch** [comment]
@userzimmermann, indeed. I must have overlooked it ;)

**Comment by thisch** [comment]
@userzimmermann, @jayvdb I hope that the commit body is fine now.

**Comment by thisch** [comment]
Should I update the pytest version in test-requirements.txt as well?

**Comment by userzimmermann** [comment]
@thisch Yes, commit message body is fine now :) And yes, please update `pytest` version in `test-requirements`

**Comment by userzimmermann** [comment]
@thisch @jayvdb @sims1253 @supergr35 Now we only have a CircleCI issue left, regarding Sphinx:

```
sphinx-build -b html -d _build/doctrees  -W . _build/html
Traceback (most recent call last):
  File "/home/ubuntu/.pyenv/versions/3.5.1/bin/sphinx-build", line 11, in <module>
    sys.exit(main())
  File "/home/ubuntu/.pyenv/versions/3.5.1/lib/python3.5/site-packages/sphinx/__init__.py", line 66, in main
    sys.exit(build_main(argv))
  File "/home/ubuntu/.pyenv/versions/3.5.1/lib/python3.5/site-packages/sphinx/__init__.py", line 76, in build_main
    from sphinx import cmdline
  File "/home/ubuntu/.pyenv/versions/3.5.1/lib/python3.5/site-packages/sphinx/cmdline.py", line 23, in <module>
    from sphinx.application import Sphinx
  File "/home/ubuntu/.pyenv/versions/3.5.1/lib/python3.5/site-packages/sphinx/application.py", line 36, in <module>
    from sphinx.domains.std import GenericObject, Target, StandardDomain
  File "/home/ubuntu/.pyenv/versions/3.5.1/lib/python3.5/site-packages/sphinx/domains/std.py", line 27, in <module>
    from sphinx.util.compat import Directive
  File "/home/ubuntu/.pyenv/versions/3.5.1/lib/python3.5/site-packages/sphinx/util/compat.py", line 19, in <module>
    docutils_version = tuple(int(x) for x in _du_version.split('.')[:2])
  File "/home/ubuntu/.pyenv/versions/3.5.1/lib/python3.5/site-packages/sphinx/util/compat.py", line 19, in <genexpr>
    docutils_version = tuple(int(x) for x in _du_version.split('.')[:2])
ValueError: invalid literal for int() with base 10: '14rc1'
```

Sphinx doesn't work with latest **rc** release of `docutils`. I raised on issue for that: https://github.com/sphinx-doc/sphinx/issues/3846

Does this currently also happen in other PRs?

cc @sils

**Comment by jayvdb** [comment]
> Does this currently also happen in other PRs?

Only PRs which have not rebased to get the workaround.

**Comment by jayvdb** [comment]
@thisch, you need to rebase your PR.

File: tests/bearlib/languages/LanguageTest.py [hunk 2]
```diff
@@ -21,9 +23,11 @@ def tearDown(self):
         pass
 
     def test_invalid_attribute(self):
-        with self.assertRaisesRegexp(AttributeError, 'not a valid attribute'):
+        with pytest.raises(AttributeError) as excinfo:
```
Comment (jayvdb) [inline tests/bearlib/languages/LanguageTest.py]: No, the deprecation warning doesnt say to do this.

We have been intentionally removing all `pytest.raises` (https://github.com/coala/coala/issues/3808), and ideally replacing them with `assertRaisesRegex` per https://github.com/coala/coala/issues/3374 which means we can prevent using the less precise `assertRaises` (https://github.com/coala/coala/issues/3373).

File: tests/results/result_actions/ApplyPatchActionTest.py [hunk 3]
```diff
@@ -115,7 +115,8 @@ def test_apply_rename(self):
                       file_diff_dict)
             self.assertFalse(isfile(f_a+'.renamed.orig'))
 
-            file_dict = {f_a+'.renamed': open(f_a+'.renamed').readlines()}
+            with open(f_a+'.renamed') as fh:
```
Comment (jayvdb) [inline tests/results/result_actions/ApplyPatchActionTest.py]: this one looks different from the other ones ... it should be a separate PR or separate commit in this PR.

Comment (thisch) [inline tests/results/result_actions/ApplyPatchActionTest.py]: This needs to be fixed in this PR, since otherwise AFAIU the tests would fail due to the new update of pytest to version 3.1. Why do you think that this fix is different from the other ones? (I also fixed a ResourceWarning in Beartest.py)
```
tests/results/result_actions/ApplyPatchActionTest.py::ApplyPatchActionTest::test_apply_rename
  /home/thomas/gitrepos/coala/tests/results/result_actions/ApplyPatchActionTest.py:118: ResourceWarning: unclosed file <_io.TextIOWrapper name='/tmp/tmpbu9i6owm.renamed' mode='r' encoding='UTF-8'>
    file_dict = {f_a+'.renamed': open(f_a+'.renamed').readlines()}
```

Comment (userzimmermann) [inline tests/results/result_actions/ApplyPatchActionTest.py]: @thisch Just put all `ResourceWarning` fixes in one commit, and all `DeprecationWarning` fixes in one commit... Just to get this PR done ;)

**Comment by userzimmermann** [comment]
@thisch Seems like `coalib` will stay trapped in `unittest` 😢 Well... There are worse things ;P As long as the tests do their job...

File: tests/results/result_actions/ApplyPatchActionTest.py (hunk 3 above)
Comment (jayvdb) [inline tests/results/result_actions/ApplyPatchActionTest.py]: Yes, grouping them together into separate commits would be my preference.  But meh.

Comment (jayvdb) [inline tests/results/result_actions/ApplyPatchActionTest.py]: Ideally also the commit message is improved, as none of these are strictly pytest related, as they occur under other test runners as well, and the regex problem actually happens in production code, not just under a test runner.

**Comment by userzimmermann** [comment]
ack https://github.com/coala/coala/pull/4276/commits/3751d78273c3a3cd4859f90eb88093047d60fe3e

**Comment by userzimmermann** [comment]
@rultor merge

**Comment by userzimmermann** [comment]
@thisch rebase :/

**Comment by thisch** [comment]
arrrghh. I'll do it in the evening

**Comment by thisch** [comment]
@userzimmermann, do you know why rultor doesn't automatically perform a rebase if there are no conflicts? The reason why I'm asking is that I did a ``git rebase origin/master``, which worked fine.

**Comment by Makman2** [comment]
Because tests may fail again after a rebase. It could be that new changes from master don't play well with your commits now^^ That's why you have to rebase manually and we let the complete CI run over your stuff again :+1:

**Comment by thisch** [comment]
Okay that  makes sense. Thx for your answer @Makman2

**Comment by thisch** [comment]
Okay that  makes sense. Thx for your answer @Makman2

**Comment by jayvdb** [comment]
Could you rebase again, and then ping a maintainer on gitter to make sure it gets merged before anything else ;-)

**Comment by jayvdb** [comment]
ack 6fa7c4e

**Comment by jayvdb** [comment]
@rultor merge

**Comment by thisch** [comment]
finally, @jayvdb and @userzimmermann thx for reviewing this!

**Comment by userzimmermann** [comment]
Thanks @thisch for removing those warnings :)
